Build fails if default-features is false

## Description

Build fails if default features if off

## Meta

Rust version: 1.37.0
Specs version / commit: 0.15.1
Operating system: Windows 10

## Reproduction
1. create a new project with ```cargo new test_project```
2. in ```cargo.toml``` write ```specs = {version = "0.15.1", default-features = false }``` in the dependencies
3. type ```cargo check``` or ```cargo build``` and wait

Error: 
```error[E0433]: failed to resolve: could not find `rayon` in `{{root}}`
 --> C:\Users\DUDDINO\.cargo\registry\src\github.com-1ecc6299db9ec823\shred-0.9.3\src\dispatch\dispatcher.rs:7:56
  |
7 | pub type ThreadPoolWrapper = Option<::std::sync::Arc<::rayon::ThreadPool>>;
  |                                                        ^^^^^ could not find `rayon` in `{{root}}`

error[E0609]: no field `thread_pool` on type `dispatch::builder::DispatcherBuilder<'a, 'b>`
   --> C:\Users\DUDDINO\.cargo\registry\src\github.com-1ecc6299db9ec823\shred-0.9.3\src\dispatch\builder.rs:238:28
    |
238 |         dispatcher_builder.thread_pool = self.thread_pool.clone();
    |                            ^^^^^^^^^^^ help: a field with a similar name exists: `thread_local`

error[E0609]: no field `thread_pool` on type `&mut dispatch::builder::DispatcherBuilder<'a, 'b>`
   --> C:\Users\DUDDINO\.cargo\registry\src\github.com-1ecc6299db9ec823\shred-0.9.3\src\dispatch\builder.rs:238:47
    |
238 |         dispatcher_builder.thread_pool = self.thread_pool.clone();
    |                                               ^^^^^^^^^^^

error: aborting due to 3 previous errors

Some errors have detailed explanations: E0433, E0609.
For more information about an error, try `rustc --explain E0433`.
error: Could not compile `shred`.
```

--- a/examples/cluster_bomb.rs
+++ b/examples/cluster_bomb.rs
@@ -1,9 +1,6 @@
-extern crate rand;
-extern crate rayon;
-extern crate specs;
-
 use rand::prelude::*;
 
+#[cfg(feature = "parallel")]
 use rayon::iter::ParallelIterator;
 
 use specs::{prelude::*, storage::HashMapStorage, WorldExt};
@@ -57,30 +54,41 @@ impl<'a> System<'a> for ClusterBombSystem {
         use rand::distributions::Uniform;
 
         let durability_range = Uniform::new(10, 20);
-        // Join components in potentially parallel way using rayon.
-        (&entities, &mut bombs, &positions)
-            .par_join()
-            .for_each(|(entity, bomb, position)| {
-                let mut rng = rand::thread_rng();
+        let update_position = |(entity, bomb, position): (Entity, &mut ClusterBomb, &Pos)| {
+            let mut rng = rand::thread_rng();
 
-                if bomb.fuse == 0 {
-                    let _ = entities.delete(entity);
-                    for _ in 0..9 {
-                        let shrapnel = entities.create();
-                        updater.insert(
-                            shrapnel,
-                            Shrapnel {
-                                durability: durability_range.sample(&mut rng),
-                            },
-                        );
-                        updater.insert(shrapnel, position.clone());
-                        let angle: f32 = rng.gen::<f32>() * TAU;
-                        updater.insert(shrapnel, Vel(angle.sin(), angle.cos()));
-                    }
-                } else {
-                    bomb.fuse -= 1;
+            if bomb.fuse == 0 {
+                let _ = entities.delete(entity);
+                for _ in 0..9 {
+                    let shrapnel = entities.create();
+                    updater.insert(
+                        shrapnel,
+                        Shrapnel {
+                            durability: durability_range.sample(&mut rng),
+                        },
+                    );
+                    updater.insert(shrapnel, position.clone());
+                    let angle: f32 = rng.gen::<f32>() * TAU;
+                    updater.insert(shrapnel, Vel(angle.sin(), angle.cos()));
                 }
-            });
+            } else {
+                bomb.fuse -= 1;
+            }
+        };
+
+        // Join components in potentially parallel way using rayon.
+        #[cfg(not(feature = "parallel"))]
+        {
+            (&entities, &mut bombs, &positions)
+                .join()
+                .for_each(update_position);
+        }
+        #[cfg(feature = "parallel")]
+        {
+            (&entities, &mut bombs, &positions)
+                .par_join()
+                .for_each(update_position);
+        }
     }
 }
 
@@ -89,6 +97,12 @@ impl<'a> System<'a> for PhysicsSystem {
     type SystemData = (WriteStorage<'a, Pos>, ReadStorage<'a, Vel>);
 
     fn run(&mut self, (mut pos, vel): Self::SystemData) {
+        #[cfg(not(feature = "parallel"))]
+        (&mut pos, &vel).join().for_each(|(pos, vel)| {
+            pos.0 += vel.0;
+            pos.1 += vel.1;
+        });
+        #[cfg(feature = "parallel")]
         (&mut pos, &vel).par_join().for_each(|(pos, vel)| {
             pos.0 += vel.0;
             pos.1 += vel.1;
@@ -101,6 +115,18 @@ impl<'a> System<'a> for ShrapnelSystem {
     type SystemData = (Entities<'a>, WriteStorage<'a, Shrapnel>);
 
     fn run(&mut self, (entities, mut shrapnels): Self::SystemData) {
+        #[cfg(not(feature = "parallel"))]
+        (&entities, &mut shrapnels)
+            .join()
+            .for_each(|(entity, shrapnel)| {
+                if shrapnel.durability == 0 {
+                    let _ = entities.delete(entity);
+                } else {
+                    shrapnel.durability -= 1;
+                }
+            });
+
+        #[cfg(feature = "parallel")]
         (&entities, &mut shrapnels)
             .par_join()
             .for_each(|(entity, shrapnel)| {

--- a/examples/full.rs
+++ b/examples/full.rs
@@ -1,6 +1,3 @@
-extern crate rayon;
-extern crate specs;
-
 use specs::{prelude::*, storage::HashMapStorage};
 
 // -- Components --
@@ -161,8 +158,10 @@ impl<'a> System<'a> for SysStoreMax {
     }
 }
 
+#[cfg(feature = "parallel")]
 struct JoinParallel;
 
+#[cfg(feature = "parallel")]
 impl<'a> System<'a> for JoinParallel {
     type SystemData = (
         ReadStorage<'a, CompBool>,
@@ -220,13 +219,19 @@ fn main() {
     // "check_positive" depends on  "print_bool" for example,
     // because we want to print the components before executing
     // `SysCheckPositive`.
-    let mut dispatcher = DispatcherBuilder::new()
+    let mut dispatcher_builder = DispatcherBuilder::new()
         .with(SysPrintBool, "print_bool", &[])
         .with(SysCheckPositive, "check_positive", &["print_bool"])
         .with(SysStoreMax::new(), "store_max", &["check_positive"])
         .with(SysSpawn::new(), "spawn", &[])
-        .with(SysPrintBool, "print_bool2", &["check_positive"])
-        .with(JoinParallel, "join_par", &[])
+        .with(SysPrintBool, "print_bool2", &["check_positive"]);
+
+    #[cfg(feature = "parallel")]
+    {
+        dispatcher_builder = dispatcher_builder.with(JoinParallel, "join_par", &[])
+    }
+
+    let mut dispatcher = dispatcher_builder
         .with_barrier() // we want to make sure all systems finished before running the last one
         .with(AddIntToFloat, "add_float_int", &[])
         .build();

--- a/examples/full.rs
+++ b/examples/full.rs
@@ -199,7 +199,8 @@ impl<'a> System<'a> for AddIntToFloat {
         }
 
         // An alternative way to write this out:
-        // (note that `entities` is just another system data of type `Èntities<'a>`)
+        // (note that `entities` is just another system data of type
+        // `Entities<'a>`)
         //
         // ```
         // for (entity, f) in (&entities, &comp_float).join() {

--- a/specs-derive/src/impl_saveload.rs
+++ b/specs-derive/src/impl_saveload.rs
@@ -60,8 +60,9 @@ pub fn impl_saveload(ast: &mut DeriveInput) -> TokenStream {
     let mut impl_generics = ast.generics.clone();
     impl_generics.params.push(parse_quote!(MA));
     let (impl_generics, saveload_ty_generics, where_clause) = impl_generics.split_for_impl();
-    let (_, ty_generics, _) = ast.generics.split_for_impl(); // We don't want the type generics we just made
-                                                             // because they have MA which our normal type doesn't have
+    // We don't want the type generics we just made because they have MA which our
+    // normal type doesn't have
+    let (_, ty_generics, _) = ast.generics.split_for_impl();
 
     let type_def = derive.type_def;
     let saveload_name = derive.saveload_name;

--- a/src/join/mod.rs
+++ b/src/join/mod.rs
@@ -291,14 +291,11 @@ where
     }
 }
 
-// SAFETY: This is safe as long as `T` implements `ParJoin` safely.  `MaybeJoin` relies on `T as
-// Join` for all storage access and safely wraps the inner `Join` API, so it should also be able to
-// implement `ParJoin`.
+// SAFETY: This is safe as long as `T` implements `ParJoin` safely.  `MaybeJoin`
+// relies on `T as Join` for all storage access and safely wraps the inner
+// `Join` API, so it should also be able to implement `ParJoin`.
 #[cfg(feature = "parallel")]
-unsafe impl<T> ParJoin for MaybeJoin<T>
-where
-    T: ParJoin,
-{}
+unsafe impl<T> ParJoin for MaybeJoin<T> where T: ParJoin {}
 
 /// `JoinIter` is an `Iterator` over a group of `Storages`.
 #[must_use]

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -224,7 +224,7 @@ pub use crate::{
     changeset::ChangeSet,
     join::Join,
     storage::{
-        DenseVecStorage, DefaultVecStorage, FlaggedStorage, HashMapStorage, NullStorage,
+        DefaultVecStorage, DenseVecStorage, FlaggedStorage, HashMapStorage, NullStorage,
         ReadStorage, Storage, Tracked, VecStorage, WriteStorage,
     },
     world::{Builder, Component, Entities, Entity, EntityBuilder, LazyUpdate, WorldExt},

--- a/src/prelude.rs
+++ b/src/prelude.rs
@@ -20,7 +20,7 @@ pub use shred::AsyncDispatcher;
 pub use crate::{
     changeset::ChangeSet,
     storage::{
-        ComponentEvent, DenseVecStorage, DefaultVecStorage, FlaggedStorage, HashMapStorage,
+        ComponentEvent, DefaultVecStorage, DenseVecStorage, FlaggedStorage, HashMapStorage,
         NullStorage, ReadStorage, Storage, Tracked, VecStorage, WriteStorage,
     },
     world::{Builder, Component, Entities, Entity, EntityBuilder, LazyUpdate, WorldExt},

--- a/src/saveload/marker.rs
+++ b/src/saveload/marker.rs
@@ -1,9 +1,10 @@
 //! Provides `Marker` and `MarkerAllocator` traits
 
-use std::{collections::HashMap,
+use std::{
+    collections::HashMap,
     fmt::{self, Debug},
     hash::{Hash, Hasher},
-    marker::PhantomData
+    marker::PhantomData,
 };
 
 use serde::{de::DeserializeOwned, Deserialize, Serialize};
@@ -429,10 +430,10 @@ pub struct SimpleMarkerAllocator<T: ?Sized> {
 impl<T: ?Sized> Debug for SimpleMarkerAllocator<T> {
     fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
         fmt.debug_struct("SimpleMarkerAllocator")
-           .field("index", &self.index)
-           .field("mapping", &self.mapping)
-           .field("_phantom_data", &self._phantom_data)
-           .finish()
+            .field("index", &self.index)
+            .field("mapping", &self.mapping)
+            .field("_phantom_data", &self._phantom_data)
+            .finish()
     }
 }
 

--- a/src/storage/data.rs
+++ b/src/storage/data.rs
@@ -67,12 +67,8 @@ use crate::{
 /// # struct Vel; impl Component for Vel { type Storage = VecStorage<Self>; }
 /// #
 /// # let mut world = World::new(); world.register::<Pos>(); world.register::<Vel>();
-/// let entity1 = world.create_entity()
-///     .with(Pos)
-///     .build();
-/// let entity2 = world.create_entity()
-///     .with(Vel)
-///     .build();
+/// let entity1 = world.create_entity().with(Pos).build();
+/// let entity2 = world.create_entity().with(Vel).build();
 ///
 /// # let pos_storage = world.read_storage::<Pos>();
 /// # let vel_storage = world.read_storage::<Vel>();
@@ -172,9 +168,7 @@ where
 /// # struct Pos(f32); impl Component for Pos { type Storage = VecStorage<Self>; }
 /// #
 /// # let mut world = World::new(); world.register::<Pos>();
-/// let entity = world.create_entity()
-///     .with(Pos(2.0))
-///     .build();
+/// let entity = world.create_entity().with(Pos(2.0)).build();
 /// # let mut pos_storage = world.write_storage::<Pos>();
 ///
 /// assert_eq!(pos_storage.get_mut(entity), Some(&mut Pos(2.0)));
@@ -197,9 +191,7 @@ where
 /// # struct Pos(f32); impl Component for Pos { type Storage = VecStorage<Self>; }
 /// #
 /// # let mut world = World::new(); world.register::<Pos>();
-/// let entity = world.create_entity()
-///     .with(Pos(0.1))
-///     .build();
+/// let entity = world.create_entity().with(Pos(0.1)).build();
 /// # let mut pos_storage = world.write_storage::<Pos>();
 ///
 /// if let Ok(Some(p)) = pos_storage.insert(entity, Pos(4.0)) {

--- a/src/storage/mod.rs
+++ b/src/storage/mod.rs
@@ -9,7 +9,9 @@ pub use self::{
         ImmutableParallelRestriction, MutableParallelRestriction, RestrictedStorage,
         SequentialRestriction,
     },
-    storages::{BTreeStorage, DenseVecStorage, DefaultVecStorage, HashMapStorage, NullStorage, VecStorage},
+    storages::{
+        BTreeStorage, DefaultVecStorage, DenseVecStorage, HashMapStorage, NullStorage, VecStorage,
+    },
     track::{ComponentEvent, Tracked},
 };
 
@@ -271,10 +273,10 @@ where
 }
 
 impl<'e, T, D> Storage<'e, T, D>
-    where
-        T: Component,
-        D: Deref<Target = MaskedStorage<T>>,
-        T::Storage: SliceAccess<T>
+where
+    T: Component,
+    D: Deref<Target = MaskedStorage<T>>,
+    T::Storage: SliceAccess<T>,
 {
     /// Returns the component data as a slice.
     ///
@@ -286,10 +288,10 @@ impl<'e, T, D> Storage<'e, T, D>
 }
 
 impl<'e, T, D> Storage<'e, T, D>
-    where
-        T: Component,
-        D: DerefMut<Target = MaskedStorage<T>>,
-        T::Storage: SliceAccess<T>
+where
+    T: Component,
+    D: DerefMut<Target = MaskedStorage<T>>,
+    T::Storage: SliceAccess<T>,
 {
     /// Returns the component data as a slice.
     ///

--- a/src/storage/mod.rs
+++ b/src/storage/mod.rs
@@ -566,6 +566,7 @@ pub trait UnprotectedStorage<T>: TryDefault {
 }
 
 #[cfg(test)]
+#[cfg(feature = "parallel")]
 mod tests_inline {
 
     use crate::{

--- a/src/storage/storages.rs
+++ b/src/storage/storages.rs
@@ -1,7 +1,6 @@
 //! Different types of storages you can use for your components.
 
-use std::collections::BTreeMap;
-use std::mem::MaybeUninit;
+use std::{collections::BTreeMap, mem::MaybeUninit};
 
 use hashbrown::HashMap;
 use hibitset::BitSetLike;
@@ -130,8 +129,8 @@ impl<T> SliceAccess<T> for DenseVecStorage<T> {
 
     /// Returns a slice of all the components in this storage.
     ///
-    /// Indices inside the slice do not correspond to anything in particular, and
-    /// especially do not correspond with entity IDs.
+    /// Indices inside the slice do not correspond to anything in particular,
+    /// and especially do not correspond with entity IDs.
     #[inline]
     fn as_slice(&self) -> &[Self::Element] {
         self.data.as_slice()
@@ -139,8 +138,8 @@ impl<T> SliceAccess<T> for DenseVecStorage<T> {
 
     /// Returns a mutable slice of all the components in this storage.
     ///
-    /// Indices inside the slice do not correspond to anything in particular, and
-    /// especially do not correspond with entity IDs.
+    /// Indices inside the slice do not correspond to anything in particular,
+    /// and especially do not correspond with entity IDs.
     #[inline]
     fn as_mut_slice(&mut self) -> &mut [Self::Element] {
         self.data.as_mut_slice()
@@ -172,7 +171,10 @@ impl<T> UnprotectedStorage<T> for DenseVecStorage<T> {
             self.data_id.reserve(delta);
             self.data_id.set_len(id + 1);
         }
-        self.data_id.get_unchecked_mut(id).as_mut_ptr().write(self.data.len() as Index);
+        self.data_id
+            .get_unchecked_mut(id)
+            .as_mut_ptr()
+            .write(self.data.len() as Index);
         self.entity_id.push(id as Index);
         self.data.push(v);
     }
@@ -180,7 +182,10 @@ impl<T> UnprotectedStorage<T> for DenseVecStorage<T> {
     unsafe fn remove(&mut self, id: Index) -> T {
         let did = self.data_id.get_unchecked(id as usize).assume_init();
         let last = *self.entity_id.last().unwrap();
-        self.data_id.get_unchecked_mut(last as usize).as_mut_ptr().write(did);
+        self.data_id
+            .get_unchecked_mut(last as usize)
+            .as_mut_ptr()
+            .write(did);
         self.entity_id.swap_remove(did as usize);
         self.data.swap_remove(did as usize)
     }
@@ -264,8 +269,8 @@ impl<T> SliceAccess<T> for VecStorage<T> {
 
 impl<T> UnprotectedStorage<T> for VecStorage<T> {
     unsafe fn clean<B>(&mut self, has: B)
-        where
-            B: BitSetLike,
+    where
+        B: BitSetLike,
     {
         use std::ptr;
         for (i, v) in self.0.iter_mut().enumerate() {
@@ -321,10 +326,13 @@ impl<T> Default for DefaultVecStorage<T> {
     }
 }
 
-impl<T> UnprotectedStorage<T> for DefaultVecStorage<T> where T: Default {
+impl<T> UnprotectedStorage<T> for DefaultVecStorage<T>
+where
+    T: Default,
+{
     unsafe fn clean<B>(&mut self, _has: B)
-        where
-            B: BitSetLike,
+    where
+        B: BitSetLike,
     {
         self.0.clear();
     }

--- a/src/storage/tests.rs
+++ b/src/storage/tests.rs
@@ -714,6 +714,7 @@ mod test {
     }
 
     #[test]
+    #[cfg(feature = "parallel")]
     fn par_restricted_storage() {
         use crate::join::ParJoin;
         use rayon::iter::ParallelIterator;
@@ -864,6 +865,7 @@ mod test {
     }
 
     #[test]
+    #[cfg(feature = "parallel")]
     fn par_storage_mask() {
         use crate::join::ParJoin;
         use rayon::iter::ParallelIterator;

--- a/src/storage/tests.rs
+++ b/src/storage/tests.rs
@@ -229,7 +229,9 @@ mod test {
     #[derive(PartialEq, Eq, Debug, Default)]
     struct CdefaultVec(u32);
     impl From<u32> for CdefaultVec {
-        fn from(v: u32) -> CdefaultVec { CdefaultVec(v) }
+        fn from(v: u32) -> CdefaultVec {
+            CdefaultVec(v)
+        }
     }
     impl AsMut<u32> for CdefaultVec {
         fn as_mut(&mut self) -> &mut u32 {
@@ -430,8 +432,8 @@ mod test {
     }
 
     fn test_slice_access<T: Component + From<u32> + Debug + Eq>()
-        where
-            T::Storage: Default + SliceAccess<T, Element=T>,
+    where
+        T::Storage: Default + SliceAccess<T, Element = T>,
     {
         let mut w = World::new();
         let mut s: Storage<T, _> = create(&mut w);
@@ -450,8 +452,8 @@ mod test {
     }
 
     fn test_maybeuninit_slice<T: Component + From<u32> + Debug + Eq>()
-        where
-            T::Storage: Default + SliceAccess<T, Element=MaybeUninit<T>>,
+    where
+        T::Storage: Default + SliceAccess<T, Element = MaybeUninit<T>>,
     {
         let mut w = World::new();
         let mut s: Storage<T, _> = create(&mut w);
@@ -569,25 +571,33 @@ mod test {
         let mut s: Storage<CdefaultVec, _> = create(&mut w);
 
         // insert 1 and 3 at 1 and 3
-        s.insert(Entity::new(1, Generation::new(1)), 1.into()).unwrap();
-        s.insert(Entity::new(3, Generation::new(1)), 3.into()).unwrap();
+        s.insert(Entity::new(1, Generation::new(1)), 1.into())
+            .unwrap();
+        s.insert(Entity::new(3, Generation::new(1)), 3.into())
+            .unwrap();
 
         // should contain default values at other locations
-        assert_eq!(s.as_slice(), &[
-            CdefaultVec(0),
-            CdefaultVec(1),
-            CdefaultVec(0),
-            CdefaultVec(3),
-        ]);
+        assert_eq!(
+            s.as_slice(),
+            &[
+                CdefaultVec(0),
+                CdefaultVec(1),
+                CdefaultVec(0),
+                CdefaultVec(3),
+            ]
+        );
 
         // deleting the record 3 should swap in the default but not shrink
         s.remove(Entity::new(3, Generation::new(1)));
-        assert_eq!(s.as_slice(), &[
-            CdefaultVec(0),
-            CdefaultVec(1),
-            CdefaultVec(0),
-            CdefaultVec(0),
-        ]);
+        assert_eq!(
+            s.as_slice(),
+            &[
+                CdefaultVec(0),
+                CdefaultVec(1),
+                CdefaultVec(0),
+                CdefaultVec(0),
+            ]
+        );
     }
 
     #[test]

--- a/src/world/lazy.rs
+++ b/src/world/lazy.rs
@@ -86,7 +86,7 @@ pub struct LazyUpdate {
 impl Default for LazyUpdate {
     fn default() -> Self {
         Self {
-            queue: Some(Default::default())
+            queue: Some(Default::default()),
         }
     }
 }

--- a/tests/tests.rs
+++ b/tests/tests.rs
@@ -1,6 +1,3 @@
-extern crate rayon;
-extern crate specs;
-
 use specs::{
     prelude::*,
     storage::HashMapStorage,
@@ -336,6 +333,7 @@ fn join_two_components() {
 }
 
 #[test]
+#[cfg(feature = "parallel")]
 fn par_join_two_components() {
     use std::sync::{
         atomic::{AtomicBool, Ordering},
@@ -398,6 +396,7 @@ fn par_join_two_components() {
 }
 
 #[test]
+#[cfg(feature = "parallel")]
 fn par_join_with_maybe() {
     use std::sync::{
         atomic::{AtomicBool, Ordering},
@@ -472,6 +471,7 @@ fn par_join_with_maybe() {
 }
 
 #[test]
+#[cfg(feature = "parallel")]
 fn par_join_many_entities_and_systems() {
     use rayon::iter::ParallelIterator;
     use std::sync::Mutex;

--- a/tests/tests.rs
+++ b/tests/tests.rs
@@ -431,18 +431,21 @@ fn par_join_with_maybe() {
         fn run(&mut self, (int, boolean): Self::SystemData) {
             use rayon::iter::ParallelIterator;
             let Iter(first, second, third, error) = *self;
-            (&int, boolean.maybe()).par_join().for_each(|(int, boolean)| {
-                let boolean = boolean.map(|c| c.0);
-                if !first.load(Ordering::SeqCst) && int.0 == 1 && boolean == Some(false) {
-                    first.store(true, Ordering::SeqCst);
-                } else if !second.load(Ordering::SeqCst) && int.0 == 2 && boolean == Some(true) {
-                    second.store(true, Ordering::SeqCst);
-                } else if !third.load(Ordering::SeqCst) && int.0 == 3 && boolean == None {
-                    third.store(true, Ordering::SeqCst);
-                } else {
-                    *error.lock().unwrap() = Some((int.0, boolean));
-                }
-            });
+            (&int, boolean.maybe())
+                .par_join()
+                .for_each(|(int, boolean)| {
+                    let boolean = boolean.map(|c| c.0);
+                    if !first.load(Ordering::SeqCst) && int.0 == 1 && boolean == Some(false) {
+                        first.store(true, Ordering::SeqCst);
+                    } else if !second.load(Ordering::SeqCst) && int.0 == 2 && boolean == Some(true)
+                    {
+                        second.store(true, Ordering::SeqCst);
+                    } else if !third.load(Ordering::SeqCst) && int.0 == 3 && boolean == None {
+                        third.store(true, Ordering::SeqCst);
+                    } else {
+                        *error.lock().unwrap() = Some((int.0, boolean));
+                    }
+                });
         }
     }
     let mut dispatcher = DispatcherBuilder::new()

